### DISCUSSION TITLE
Modifying pod templates with containermode.type=kubernetes

### DISCUSSION THREAD

**Discussion by nikolamarunic** [actions/actions-runner-controller#3107] [Questions] [answered]
Hello,

My current understanding of containermode.type=kubernetes is that there is essentially two pods that are spun up by the action, one being the 'runner' that communicates with GitHub and the other the 'worker' which is where the actual job will run.

My question is related to modifying the pod template for the worker (for example, if we want to add some sort of resource requests/limits/etc). I think this would also be useful i.e to inject Kubernetes secrets into the worker pods.

**Is it currently possible to modify this worker pod template?** I know we can modify the values.yml which will affect the runner pod, but it doesn't seem to apply the same fields to the other pod that is created. I see some similar discussions but don't see any official solution.

Thanks in advance!

**Comment by nikola-jokic** [comment, marked as answer]
Hey @nikolamarunic,

You can use [hook extensions](https://github.com/actions/runner-container-hooks/blob/main/docs/adrs/0096-hook-extensions.md) to customize job pods that are created by the kubernetes hook. Probably the best way to do so is to create a config map and mount it on the runner. Set the env to point to the hook extension file and the hook should apply the spec according to the ADR. :relaxed:

**Reply by nikolamarunic** [reply]
Thank you Nikola - will look into using this :)

**Reply by David9902** [reply]
@nikola-jokic 
It would be great if an example could be added to the documentation on how to add resource limits (cpu and memory) to the workflow container that is spawened by the runner.

**Reply by rteeling-evernorth** [reply]
@David9902 
For posterity, this is in the  examples folder of the runner container hooks repo here: https://github.com/actions/runner-container-hooks/blob/main/examples/extension.yaml

**Reply by jonwest** [reply]
Thanks for that @rteeling-evernorth!

I'm confused by the implementation, based on the documentation.  😅  

In [the limitations section of the README](https://github.com/actions/runner-container-hooks/blob/main/packages/k8s/README.md#limitations) it says:

> - Container actions will not have access to the services network or job container network

In the example you provided, there is [a `redis` service mentioned](https://github.com/actions/runner-container-hooks/blob/9705deeb083452f326cb790231645a2618955bfa/examples/extension.yaml#L23).

I've got a couple of questions regarding the two links above, and how they correspond with each other:
- If the container action doesn't have access to the `services` network, or the `job container` network, then how does it communicate with the `redis` service?
- If the container action _cannot_ communicate with the `redis` service, then what would the `redis` service be used for?
- Related to the second question—what would be the use case for _any_ sidecars alongside the runner if they cannot be used by the action?  Metrics, maybe?

**Reply by cgiraldo** [reply]
The `redis` service container and the `job` container will be part of the same pod, so they share the network namespace. 

You can access `redis` port from the `job` container using localhost or 127.0.0.1.

**Comment by arun291091** [comment]
This worked for me 

[3108](https://github.com/actions/actions-runner-controller/issues/3108#issuecomment-3159722348 )

**Comment by willbrownecs** [comment]
The following worked for us using DIND mode

values.yaml =>
```
...
# containerMode:
#   type: dind # allows for running docker commands/containers within the pod

template:
  spec:
    resources:
      requests:
        cpu: 250m
        memory: 512Mi
      limits:
        # cpu: 100m
        memory: 2Gi
    initContainers:
      - name: init-dind-externals
        image: ghcr.io/actions/actions-runner:latest
        command:
          ["cp", "-r", "/home/runner/externals/.", "/home/runner/tmpDir/"]
        volumeMounts:
          - name: dind-externals
            mountPath: /home/runner/tmpDir
      - name: dind
        image: docker:dind
        args:
          - dockerd
          - --host=unix:///var/run/docker.sock
          - --group=$(DOCKER_GROUP_GID)
        env:
          - name: DOCKER_GROUP_GID
            value: "123"
        securityContext:
          privileged: true
        restartPolicy: Always
        startupProbe:
          exec:
            command:
              - docker
              - info
          initialDelaySeconds: 0
          failureThreshold: 24
          periodSeconds: 5
        volumeMounts:
          - name: work
            mountPath: /home/runner/_work
          - name: dind-sock
            mountPath: /var/run
          - name: dind-externals
            mountPath: /home/runner/externals
    containers:
      - name: runner
        image: ghcr.io/actions/actions-runner:latest
        command: ["/home/runner/run.sh"]
        env:
          - name: DOCKER_HOST
            value: unix:///var/run/docker.sock
          - name: RUNNER_WAIT_FOR_DOCKER_IN_SECONDS
            value: "120"
        volumeMounts:
          - name: work
            mountPath: /home/runner/_work
          - name: dind-sock
            mountPath: /var/run
    volumes:
      - name: work
        emptyDir: {}
      - name: dind-sock
        emptyDir: {}
      - name: dind-externals
        emptyDir: {}
```
